FSW/src/util/timing.cpp

```diff
@@ -0,0 +1,237 @@
+            scienceMode.set(STANDBY_MODE);
+        }
+
+    } else if (currentMode == STANDBY_MODE){
+        // TODO: currently the only way to leave standby is through a command,
+        //       do we want some automated way to leave for testing?
+        //       it honestly might be best to leave it as having to manually start
+        //       a data window for testing
+        
+        // execute commands  
+        commandHandling();
+
+        // TODO: memory scrubbing  
+
+    } else { // mode was not recognized, this should not happen
+        Serial.println("Payload Mode not recognized (Timing Module) - Defaulting to Safe Mode");
+        scienceMode.set(SAFE_MODE);
+    }
+}
+
+/* - - - - - - smoothBuffer - - - - - - *
+ * Usage:
+ *  Apply basic smoothing to a configurable number of the most recent buffer entries
+ *  Avoids a single noisy measurement prematurely ending a window
+ *  To change the number of indices used in smoothing, change SMOOTH_IDX_COUNT in config.hpp 
+ * 
+ * Inputs:
+ *  buffer - array of photodiode voltages
+ *  bufIdx - index of most recent photodiode measurement
+ * 
+ * Outputs:
+ *  smoothVoltage - average of the previous SMOOTH_IDX_COUNT indices of buffer
+ */
+float smoothBuffer(float buffer[], int bufIdx){
+    float smoothVoltage = 0;
+    int i, idx;
+    for (i = bufIdx; i > bufIdx - SMOOTH_IDX_COUNT; i--){ // decrement to move backwards in time
+        
+        idx = i; // I am scared to toss the iterator for this loop through wrapBufferIdx
+        
+        // ensure that index is within array bounds 
+        // (i.e. starting with i=0 would throw error if we didn't wrap the index)
+        idx = wrapBufferIdx(idx); 
+
+        smoothVoltage += buffer[idx]; // sum voltages
+    }
+    return smoothVoltage / (float)SMOOTH_IDX_COUNT; // return mean
+}
+
+/* - - - - - - wrapBufferIdx - - - - - - *
+ * Usage:
+ *  Avoid indexing the photodiode voltage buffer with a negative value
+ * 
+ * Inputs:
+ *  idx - an index that may be outside of the buffer array size
+ * 
+ * Outputs:
+ *  idx - equivalent index, but guaranteed to be within buffer array size
+ */
+int wrapBufferIdx(int idx){
+    const int NUM_TRIES = 10; // max iterations (prevent infinite loop)
+
+    for (int j = 0; j < NUM_TRIES; j++){
+        if (idx < 0) {                  // wrap index if it is negative
+            idx = idx + BUFFERSIZE;     // -1 becomes BUFFERSIZE - 1
+        } else if (idx >= BUFFERSIZE) { // wrap index if it is too large
+            idx -= BUFFERSIZE;          // BUFFERSIZE + 1 becomes 1
+        } else {                        // within valid bounds for indexing buffer
+            return idx;
+        }
+    }
+    Serial.print("wrapBufferIdx failed to find valid index after ");
+    Serial.print(NUM_TRIES);
+    Serial.println(" tries.");
+    Serial.println("Defaulting to an index of 0 (Timing module)");
+    
+    // TODO: Should this return -1 instead to throw an error?
+    return 0;

```

Comment:
Do you need any return here outside of the if statement? Seems like the if statement covers any possible case? I'd have this return -1 because something is definitely fucked in the event that if block doesn't return a value. If you return zero, you may end up in a case that's quite hard to debug.

Reply:
yeah it shouldn't ever return here, I will change it to -1 so that we definitely know what went wrong